common/lib/esp32-c3-dkc02-bsc/src/wifi.rs

```diff
@@ -63,22 +64,34 @@ pub fn wifi(ssid: &str, psk: &str) -> anyhow::Result<Wifi> {
         ssid: ssid.into(),
         password: psk.into(),
         channel,
-        auth_method: auth_method,
+        auth_method,
         ..Default::default()
     }))?;
 
     info!("getting Wifi status");
 
-    let status = wifi.get_status();
-
-    if let wifi::Status(
-        ClientStatus::Started(ClientConnectionStatus::Connected(ClientIpStatus::Done(_))),
-        _,
-    ) = status
-    {
-        info!("Wifi connected!");
-    } else {
-        bail!("Unexpected Wifi status: {:?}", status);
+    let mut status = wifi.get_status();
+    // loop over the status' value until it is either connected or disconnected
+    // catch all is to make sure that at least if more states are added, the enum here
+    // remains exhaustive even though logically it might not stand up the test of time
+    while let wifi::Status(ClientStatus::Started(ref client_connection_status), _) = status {
```

Comment:
How likely is it that this block ends up in an infinite loop?

Reply:
What about using a [wait with timeout like in `rust-esp32-std-demo`](https://github.com/ivmarkov/rust-esp32-std-demo/blob/main/src/main.rs#L1353-L1371)

Reply:
@justahero I think that is a seriously important question. Given how little I know how wifi status work, I am not sure.

@SergioGasquez hmm I tried with the same code block but wifi now never connects.  It always ends up in the `map_err` in the line for `wait_status_with_timeout` here -

```
wifi.wait_status_with_timeout(Duration::from_secs(21), |status| !status.is_transitional())
        .map_err(|err| anyhow::anyhow!("Unexpected Wifi status part 1: {:?}", err))?;
```
Let me play with this a little more today. It is 7am here and I had 15 mins to try so I did.

Reply:
Ok folks, I tried @SergioGasquez 's suggestion but that does not work as-is. I tried to debug thinking maybe somewhere in the chain of `is_transitional`s there maybe something missing but I could not see anything logically incorrect. However, before the `wifi.wait_status_with_timeout(Duration...` if I add a `thread::sleep` for 2 seconds, this code works fine. This makes little sense to me since the `wifi.wait_status_with_timeout` seems like the right thing for this kind of job.

Reply:
OK not sure how big a headway this is but [here is dump of the logs of the connection using your suggestion](https://gist.github.com/amanjeev/0a752f3a4edd34577804b99e45bf6880), @SergioGasquez 

I can tell you that the wait of 20 odd seconds does not happen but I could be wrong. 

```
I (3471) esp_idf_svc::wifi: About to wait 2100s for status
I (3481) esp_idf_svc::wifi: Waiting for status done - success
I (3491) esp_idf_svc::wifi: Providing status: Status(Started(Connecting), Stopped)
I (3491) esp_idf_svc::wifi: Stopping
```
I wonder if its these lines - https://github.com/esp-rs/esp-idf-svc/blob/master/src/wifi.rs#L514-L515

Any thoughts how to proceed?

Reply:
after trying to resolve this by using a local copy of idf-svc, i realized that the dependencies were old. so then I fought cargo a bit because cargo was unhappy with the conflicts of `embedded-hal`. but i am happy to report that this is working now and with the solution provided by @SergioGasquez 

thank you!